Enable binding to CoreLib as NI image

gkhanna79 opened the pull request:
Fixes https://github.com/dotnet/coreclr/issues/10097

@jkotas Initial change - still testing.

jkotas commented:
Looks good to me so far.

gkhanna79 commented:
@jyoungyun @hqueue Can you help me understand why the Ubuntu Arm test run is failing? I do not have infrastructure to run the tests locally, so it will be helpful if you can point me in the right direction.

gkhanna79 commented:
@dotnet-bot test Ubuntu arm Cross Release Build  please

gkhanna79 commented on src/binder/assemblybinder.cpp:
```diff
@@ -685,19 +685,23 @@ namespace BINDER_SPACE
+#if defined(FEATURE_PAL) && !defined(_TARGET_AMD64_)
+        // Native image for core library only exists for X64 on non-Windows. As other architectures
+        // bring them up, we should update this condition.
+        fBindToNativeImage = false;
```
@jkotas I had to add this for Linux-Arm builds. My observation is that if the first call to GetAssembly below attempts to bind to SPC.dll (which is IL binary on Linux-Arm since it is not crossgenned yet), it likely results in a failure and thus, CLR startup fails.

Setting this to false makes us attempt to use the inner GetAssembly call that attempts to bind to IL assembly with fExplicitBindToNativeImage as FALSE. And this, approach appears to work (and is what was prior to my change as well since spc.ni.dll didnt exist for Arm).

I dont have a way to setup a repro for running Linux Arm tests, so I adopted setting the flag to false for now though I dont really like it though.

Do you have ideas on why the failure maybe happening since the failure case (bind to IL SPC.dll as with fExplicitBindToNativeImage  as TRUE) works on Windows?

gkhanna79 replied:
CC @janvorli as well incase he has any ideas on this.

jkotas replied:
I guess that the PAL LoadLibrary simulator chokes on the IL image.

janvorli replied:
> I dont have a way to setup a repro for running Linux Arm tests

@gkhanna79, do you mean that you don't have a Linux ARM device for testing or is the problem not reproing always?

gkhanna79 replied:
The former @janvorli. If you can help investigate on a device (by setting the bool to be true or removing the ifdef I added) and figure out why it fails, that will be great.

janvorli replied:
@gkhanna79 sure, I can try that tomorrow.

gkhanna79 commented:
You mean it chokes on IL images whose PEImage is marked as NI, since it works fine when operating on IL images as IL?

Other than this, can you please review the change?


Thanks,
Gaurav
________________________________
From: Jan Kotas <notifications@github.com>
Sent: Apr 19, 2017 9:47 PM
To: dotnet/coreclr
Cc: Gaurav Khanna (CLR); Author
Subject: Re: [dotnet/coreclr] [WIP] Enable binding to CoreLib as NI image (#11040)


@jkotas commented on this pull request.

________________________________

In src/binder/assemblybinder.cpp<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcoreclr%2Fpull%2F11040%23discussion_r112364886&data=02%7C01%7CGaurav.Khanna%40microsoft.com%7C74de153f38604865acfb08d487a85b09%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636282604527406171&sdata=iUokzYhIvxPiGCSp1cPnhBxKRiGfUVOO5u55EEtb4dI%3D&reserved=0>:

>          sCoreLib = sCoreLibDir;
-        sCoreLib.Append(CoreLibName_NI_W);
+        sCoreLib.Append(CoreLibName_IL_W);
+#if defined(FEATURE_PAL) && !defined(_TARGET_AMD64_)
+        // Native image for core library only exists for X64 on non-Windows. As other architectures
+        // bring them up, we should update this condition.
+        fBindToNativeImage = false;


I guess that the PAL LoadLibrary simulator chokes on the IL image.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcoreclr%2Fpull%2F11040%23discussion_r112364886&data=02%7C01%7CGaurav.Khanna%40microsoft.com%7C74de153f38604865acfb08d487a85b09%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636282604527406171&sdata=iUokzYhIvxPiGCSp1cPnhBxKRiGfUVOO5u55EEtb4dI%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJA3wXdKzIYB5szDY17iJWbS8L-RI8U0ks5rxuNhgaJpZM4M_4D4&data=02%7C01%7CGaurav.Khanna%40microsoft.com%7C74de153f38604865acfb08d487a85b09%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636282604527406171&sdata=71HOrLE62nqLkDgEL7FCczki9xXh3PjHWl%2FxgrAX3Hk%3D&reserved=0>.

jkotas commented on src/binder/assemblybinder.cpp:
```diff
@@ -685,19 +685,23 @@ namespace BINDER_SPACE
-        sCoreLib.Append(CoreLibName_NI_W);
+        sCoreLib.Append(CoreLibName_IL_W);
+#if defined(FEATURE_PAL) && !defined(_TARGET_AMD64_)
+        // Native image for core library only exists for X64 on non-Windows. As other architectures
```
The only real difference that fExplicitBindToNativeImage makes is whether we throw when LoadLibrary fails, or whether we take the fallback path. It is done here: https://github.com/dotnet/coreclr/blob/3c61ffd676603bb14d4e815e1e249f816d4cbc74/src/vm/peimage.cpp#L918. I think we should be able to remove all this if we start unconditionally passing in `bThrowOnError = FALSE` here.

jkotas replied:
Nice to have cleanup would be to remove `m_bIsTrustedNativeImage` and related flags from everywhere.

gkhanna79 replied:
Thanks @jkotas - this is really helpful. If GetAssembly would always pass fExplicitBindToNativeImage as FALSE, would the resulting bind still have the precompiled native code kick-in OR will it be treated purely as IL and triggering JIT?

jkotas replied:
> the resulting bind still have the precompiled native code kick-in 

There should be nothing fundamental preventing it. There may be some other random check that derails it that we need to remove.

gkhanna79 replied:
For now, I am keeping the change as is and have rebased it. @jkotas PTAL.

jkotas replied:
I suspect that there will be other situations that run with IL CoreLib where this will break. I cannot think about any real one though.

So LGTM

jkotas replied:
I suspect that there will be other situations that run with IL CoreLib where this will break. I cannot think about any real one though.

So LGTM

gkhanna79 commented:
@janvorli I rebased against PAL alignment change from @sdmaclea and updated my PR to always bind and yet it fails on Arm - can you help me understand this?

janvorli commented:
@gkhanna79 sure, I'll take a look.

janvorli commented:
@gkhanna79 The issue is due to an invalid address passed to the mmap function. The errno was 22: Invalid argument.

```C++
   2253     pvBaseAddress = mmap(static_cast<char *>(addr) - adjust, len + adjust, prot, flags, fd, offset - adjust);
   2254     if (MAP_FAILED == pvBaseAddress)
   2255     {
-> 2256         ERROR_(LOADER)( "mmap failed with code %d: %s.\n", errno, strerror( errno ) );
   2257         palError = FILEGetLastErrorFromErrno();
   2258     }
   2259     else
```
The `addr` is 0x10002000 and `adjust` is 512, `len` is 0x002b8000, `prot` is 5, `flags` is 0x12 and `offset` is 512.
The problem is that `addr - adjust` is not page aligned and mmap doesn't accept that. I've verified that it is the problem by hand-modifying the addr to 0x10002200 right before the mmap call so that it would get an aligned address and it passed.

@sdmaclea could you please look into fixing that?

sdmaclea commented:
@janvorli @gkhanna79 Can someone provide some more context.  I am not sure exactly what is trying to be accomplished and why my patch was relevant.  

Are we trying to load an Elf file?  If so, can someone post an objdump of the section headers?

There was a late change to my PR which effectively disabled removing the file alignment restriction from arm crossgen files.  It is trivial to reenable.  I can provide details when I get to the office.

janvorli commented:
@sdmaclea the issue is that with this change, we are trying to load an IL image. Before your change, the MAPMapPEFile was failing because of the optional header having alignment of 0x200 in the piece of code that you have removed:
```C++
   //this code requires that the file alignment be the same as the page alignment		
   if (ntHeader.OptionalHeader.FileAlignment < VIRTUAL_PAGE_SIZE)		
   {		
       ERROR_(LOADER)( "Optional header file alignment is bad\n" );		
       palError = ERROR_INVALID_PARAMETER;		
       goto done;		
   }
```
So we were hoping that your change would allow us to succeed loading that image because the file alignment requirement is supposed to be removed. But we started to get error in the MAPmmapAndRecord instead due the the condition mentioned in my previous comment.

gkhanna79 commented:
Specifically, we are trying to load an image as crossgen compiled native image (NI), whether or not it has native code in it. On X64, we can crossgen System.Private.CoreLib.dll so we generate its NI and rename to S.P.C.dll when placing in nuget package. For other arch (e..g Linux arm) we cannot do so yet, so we end up placing IL version of the library.

The binder can make out the difference between the two and treat them appropriately. 

This works on Windows but is failing on Linux Arm.

Hope this gives enough context @sdmaclea.

sdmaclea commented:
@gkhanna79 mmap makes a fundamental assumption.  It loads the physical file is loaded into physical memory with assert((fileoffset & (PAGE_SIZE-1)) == (physical address & (PAGE_SIZE-1)));

It then allows you to map any physical page of the file to one or more virtual address.  With the property, 
assert((virtual address & (PAGE_SIZE-1)) == (physical address & (PAGE_SIZE-1))).  This leads to the derived assert((fileoffset & (PAGE_SIZE-1)) == (virtual address & (PAGE_SIZE-1))); 

The secondary constraint is that the permissions are set at a page granularity.

So my patch was trying to remove the artificial requirement that the sections be aligned to page boundaries.  Leaving only the fundamental requirements for using mmap.  So for arm64, to support 64k pages, I adjusted the zapwriter to pad the VA by PAL_MAX_PAGE_SIZE.  Sections are now only aligned to the PE file alignment.  Per PE spec the minimum alignment is 512 bytes.  So that is what we are using on 64bit systems.

For arm32, this padding is zero, because the file alignment is currently set to 4K.  This means every section of an arm32 crossgen file is aligned in the file to a page boundary, so padding is not necessary (assuming 4k pages on arm32).

See 
- https://github.com/dotnet/coreclr/blob/master/src/zap/zapwriter.cpp#L58-L64
- https://github.com/dotnet/coreclr/blob/master/src/zap/zapimage.cpp#L1548-L1550

So to make this work for a non-crossgen il image, you will need to either
- Fix the section alignment of the IL image so that the it meets the mmap requirements above.
- Only use mmap if the alignment requirements for using mmap are met.  Otherwise, use memcpy or some other means to locate the fileimage correctly.

gkhanna79 commented:
Thanks for the explanation @sdmaclea. 

For now, I would isolate investigating the PAL load issue for IL seperate from this to ensure I can get this baked and stabilized. @jkotas Are you good with this approach? I have updated PR to bind as IL for platforms we dont support NI generation yet.

sdmaclea commented on src/binder/assemblybinder.cpp:
```diff
@@ -689,6 +689,11 @@ namespace BINDER_SPACE
         BOOL fExplicitBindToNativeImage = (fBindToNativeImage == true)? TRUE:FALSE;
+#if defined(FEATURE_PAL) && !defined(_TARGET_AMD64_)      
+        // Non-Amd64 platforms on non-Windows do not support generating the NI image
+        // as CoreLib.dll. For those, we will bind as IL.
```
@gkhanna79 Why does this work on AMD64 *NIX?  What is the root cause?  I assume this has something to do with out mmap discussion.

gkhanna79 replied:
I suspect the PAL issue is present on AMD64 as well. However, on AMD64, SPC.dll is a real NI image and not IL image and hence, it works.

jkotas commented:
>  @jkotas Are you good with this approach? 

Yes.

gkhanna79 commented:
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test please

gkhanna79 commented:
@Petermarcu @weshaggard FYI - This is the change that will fix https://github.com/dotnet/coreclr/issues/10097 to ensure M.N.App only contains the NI images of SPC.dll.

I hope to merge this tomorrow.

Petermarcu commented:
Looks good. Are you wanting this for Preview1 as well or just in Master?

gkhanna79 commented:
I will be making this change in the master. However, this change effectively remove S.P.C.ni.dll by only having one binary in the package: S.P.C.dll:

On platforms that support NI generation, S.P.C.ni.dll is renamed to S.P.C.dll and bound as NI. For other platforms (e.g. Linux Arm or Linux X86), where cross-targeting crossgen is not available to generate the NI image, we keep the original (IL) S.P.C.dll as is.

Having a consistent name, agnostic of whether it contains NI or no, lines up with the direction of having single naming scheme where folks do not have to distinguish between NI or IL as @jkotas mentioned in https://github.com/dotnet/coreclr/pull/10101#issuecomment-285800963 and https://github.com/dotnet/coreclr/pull/10101#issuecomment-285803600.

In a manner of speaking, this will be a breaking change for those who rely on the S.P.C.ni.dll (less impactful than the renaming of mscorlib -> SPC.dll that we did around 1.0 RTM timeframe). Do you have appetite for such a change in master? I think we should take this (as opposed to removing SPC.dll from RID split M.N.App in Core-Setup) but I wanted to raise this point. Let me know what you think.

That said, @livarcocc @eerhardt Does SDK or CLI rely on SPC.ni.dll such that this will break it?

weshaggard commented:
@stephentoub @danmosemsft this will break code coverage runs in corefx because the coverage runs do not work with ni assemblies and we have traditionally deleted the ni and just used IL for those runs.

jkotas commented:
What is special about the code coverage runs that they do not work with the native image renamed to .dll?

eerhardt commented:
> Does SDK or CLI rely on SPC.ni.dll such that this will break it?

I don't know for sure.  Nothing sticks out off the top of my head, but that doesn't mean it will just work.

gkhanna79 commented:
I chatted with @Petermarcu offline and he is also supportive of bringing this change in. To minimize disruption, @weshaggard, can we make the code-coverage infra stop doing the deletion of NI?

danmosemsft commented:
Do we know of a way to make CC still work? 

If not I think at this point we can afford for CC to go dark for a while, if we have to.

weshaggard commented:
>@weshaggard, can we make the code-coverage infra stop doing the deletion of NI?

It should just work as all we do is filter oiut the *.ni files https://github.com/dotnet/corefx/blob/master/external/runtime/runtime.depproj#L93 

>What is special about the code coverage runs that they do not work with the native image renamed to .dll?

I'm not exactly sure but I believe it is because they cannot rewrite the ni's for doing the coverage runs. @stephentoub might know more of those details.

stephentoub commented:
> I'm not exactly sure but I believe it is because they cannot rewrite the ni's for doing the coverage runs.

That's my understanding as well; if the JIT isn't generating code, then the profiling APIs for instrumenting that generated code aren't invoked, and the sequence points needed by the code coverage tool aren't inserted.

jkotas commented:
The profiling APIs lets specify that you want to instrument the code and do not want to use precompiled native code. If you specify this flag and CoreLib is precompiled, we will still load it but we will ignore the native code in the precompiled image. It has worked like that since Windows Phone 8.0.

The code to make it happen is under `FEATURE_TREAT_NI_AS_MSIL_DURING_DIAGNOSTICS`  (https://github.com/dotnet/coreclr/blob/master/src/inc/switches.h#L222).

stephentoub commented:
Might be a different issue.  When I try to use the dll/pdb produced by `build.cmd release` in coreclr, code coverage fails.  If I instead use the build produced by `build.cmd windowsmscorlib release` in coreclr, code coverage succeeds.  For the former, it doesn't matter whether COMPlus_ZapDisable=1 is set, nor whether I use the System.Private.CoreLib.pdb or the System.Private.CoreLib.ni.pdb produced by the build.
